File: Tree_Postorder_Traversal/Tree_Postorder_Traversal.ts
//Tree Postorder Traversal in TypeScript

export {}

// Tree Nodes 
interface treeNodes
{
    val: number;
    parent?: treeNodes;  
    left?: treeNodes;   
    right?: treeNodes;  
}

// Defining the nodes with their values and left right children. If a value is not given it is undefined.
const node11: treeNodes = { val: 11 };
const node10: treeNodes = { val: 10 };
const node9: treeNodes = { val: 9 };
const node8: treeNodes = { val: 8 };
const node7: treeNodes = { val: 7 };
const node6: treeNodes = { val: 6, left: node10, right: node11};
const node5: treeNodes = { val: 5 };
const node4: treeNodes = { val: 4, left: node8, right: node9 };
const node3: treeNodes = { val: 3, left: node6, right: node7 };
const node2: treeNodes = { val: 2, left: node4, right: node5 };
const root:  treeNodes = { val: 1, left: node2, right: node3 };

// Construct a tree by assigning parent
function constructBinaryTree(): treeNodes
{
    root.parent = undefined;
    node2.parent = root;
    node3.parent = root;
    node4.parent = node2;
    node5.parent = node2;
    node6.parent = node3;
    node7.parent = node3;
    node8.parent = node4;
    node9.parent = node4;
    node10.parent = node6;
    node11.parent = node6;

    return root;
}

// Recursive Function to call itself on left child, then on right child and then print value of current node 
function postOrderTraversal(currentNode: treeNodes) 
{
    if (!currentNode)
        return;
    postOrderTraversal(currentNode.left); 
    postOrderTraversal(currentNode.right);
    console.log(currentNode.val + " -> ");
}

constructBinaryTree();
console.log("The postorder traversal of the binary tree is :");
postOrderTraversal(root);

/*
Output:
The postorder traversal of the binary tree is :
8 -> 
9 -> 
4 -> 
5 -> 
2 ->
10 ->
11 -> 
6 -> 
7 -> 
3 -> 
1 ->
*/

File: Tree_Postorder_Traversal/Tree_Postorder_Traversal.kt
import java.util.*

class Node<Int>(
    var key:Int,
    var left:Node<Int>? = null,
    var right:Node<Int>? = null  
)
{
    fun postorderTraversal()
    {
        left?.postorderTraversal()
        right?.postorderTraversal()
        print("$key ")
    }
}

fun main() {
    var read = Scanner(System.`in`)
    println("Enter the size of Array:")
    val arrSize = read.nextLine().toInt()
    var arr = IntArray(arrSize)
    val nodes = mutableListOf<Node<Int>>()
    println("Enter the array respresentaion of binary tree")
    
    for(i in 0 until arrSize) 
    {
        arr[i] = read.nextLine().toInt()
        nodes.add(Node(arr[i]))
    }
    
    for(i in 0..arrSize - 2)
    {
        if ((i * 2) + 1 < arrSize && arr[(i * 2) + 1] != -1)
        {
            nodes[i].left = nodes[(i * 2) + 1]
        }
	    
        if ((i * 2) + 2 < arrSize && arr[(i * 2) + 2] != -1)
        {
            nodes[i].right = nodes[(i * 2) + 2]
        }
    }
    print("Post Order traversal of tree is ")
    nodes[0].postorderTraversal()
}

/*
*Enter the size of Array:
*7
*Enter the array respresentaion of binary tree
*1
*2
*3
*4
*5
*6
*-1
*Post Order traversal of tree is 4 5 2 6 3 1
*/
